### DISCUSSION TITLE
Error when Importing openai

### DISCUSSION THREAD

**Discussion by aalrabah** [PrairieLearn/PrairieLearn#10168] [Dev]
I'm running a Docker container locally and testing some questions in PrairieLearn. I'm trying to integrate an OpenAI API call into my grading process so that a specific message pops up whenever a problem is graded in PrairieLearn. However, I've encountered several challenges and would appreciate some guidance. Here's what I've attempted so far:

1. I tried installing OpenAI in the serverFilesCourse directory using:

`docker exec -it [container_name] pip3 install --target /course/serverFilesCourse openai`

and then added this to my info.json ` "serverFilesCourse": ["sql-grader/", "grader/", "openai/"],`

This seemed to install OpenAI, but I couldn't import it in my grading scripts in serverFilesCourse/. Whenever i submit something in the local server I keep getting (No module named 'openai')

2. I'm running the PrairieLearn server with this command:
`docker run -it --rm -p 3000:3000 -v "$PWD:/course" -v "$HOME/pl_ag_jobs:/jobs" -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" -v /var/run/docker.sock:/var/run/docker.sock --platform linux/amd64 --add-host=host.docker.internal:host-gateway prairielearn/prairielearn`

**Comment by nwalters512** [comment]
Is this for an external grader? If so, you should install the package into the image via your Dockerfile. This is much easier than trying to muck around with `serverFilesCourse`.

**Reply by aalrabah** [reply]
yup that is for the external grader, I was able to add open ai to my dockerfile using the following command: 

```
FROM ubuntu:latest

# Update and install system packages
RUN apt-get update && \
    apt-get install -y python3-pip python3-dev sudo python3-venv && \
    DEBIAN_FRONTEND=noninteractive apt-get install -y mysql-server && \
    rm -rf /var/lib/apt/lists/*

# Create and activate a virtual environment
RUN python3 -m venv /opt/venv && \
    . /opt/venv/bin/activate && \
    python3 -m pip install --upgrade pip && \
    python3 -m pip install sqlparse mysql-connector-python rstr openai

# Copy configuration and set permissions
COPY my.cnf /etc/mysql/my.cnf
RUN mkdir -p /var/run/mysqld/ && chown mysql:mysql /var/run/mysqld
VOLUME /var/lib/mysql

# Ensure the virtual environment is used
ENV VIRTUAL_ENV=/opt/venv
ENV PATH="$VIRTUAL_ENV/bin:$PATH"

```

I then rebuild the docker and ran the server, an `import openai `code has been added to one of the .py files in the `serverFilesCourse`

I still run into the same issue after submitting an answer in the Prairielearn server `ModuleNotFoundError: No module named 'openai'`

**Reply by aalrabah** [reply]
@nwalters512 

So I was able to create a custome docker image and overcome the module not found error. I do have a connectivity errors though that is quite interesting. When i attempt to answer a question in prairielearn server I get the following error in my terminal:

<img width="1089" alt="Screenshot 2024-07-02 at 4 25 19 PM" src="https://github.com/PrairieLearn/PrairieLearn/assets/94386936/3192f14b-8f86-4a03-9cce-7785c1fa75eb">

Here's what I've tried:

1- Created a custom Ubuntu-based Docker image with Python and OpenAI library.
2- Set OpenAI API key as an environment variable.
3- Confirmed internet connectivity within the container (ping to google.com works).
4- Attempted to configure DNS (8.8.8.8 and 8.8.4.4) in the container.
5- Implemented retry logic with exponential backoff for API calls.

The container has internet access, but can't resolve the API's domain. I followed the exact openai documentation for creating an object for an API call. 

Has anyone encountered similar issues or have suggestions for resolving this API problem specifically for the OpenAI API in a Docker environment?

**Reply by jonatanschroeder** [reply]
Have you [enabled networking](https://prairielearn.readthedocs.io/en/latest/externalGrading/#configuring-and-enabling-external-grader-support)?

**Reply by aalrabah** [reply]
Perfect finally it worked after enabling network, I wasn't aware of that. This is helpful thank you very much for your support !!
